File: core/src/main/java/org/elkd/core/log/InMemoryLog.java
package org.elkd.core.log;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import org.elkd.core.consensus.messages.Entry;

import java.util.ArrayList;
import java.util.List;

public class InMemoryLog implements Log<Entry> {
  private static final int START_INDEX = 0;
  private List<Entry> mLogStore;
  private long mIndex;
  private long mCommitIndex;

  public InMemoryLog() {
    mLogStore = new ArrayList<>();

    /* Default Log Record - prevents having to provide a log index that
       technically does not exist e.g index = -1 */
    append(Entry.NULL_ENTRY);
  }

  @Override
  public long append(final Entry entry) {
    Preconditions.checkNotNull(entry, "entry");

    insertOrReplace((int) mIndex, entry);

    return mIndex++;
  }

  @Override
  public long append(final long index, final Entry entry) {
    Preconditions.checkState(0 <= index && index <= mIndex, "index");

    insertOrReplace((int) index, entry);
    return index;
  }

  @Override
  public Entry read(final long index) {
    try {
      Preconditions.checkState(START_INDEX <= index && index <= mIndex);
      return mLogStore.get((int) index);
    } catch (final Exception e) {
      return null;
    }
  }

  @Override
  public List<Entry> read(final long from, final long to) {
    Preconditions.checkState(START_INDEX <= from && from <= to);

    final List<Entry> subList = new ArrayList<>();
    for (int i = (int) from; i <= to; ++i) {
      final Entry entry = read(i);
      if (entry != null) {
        subList.add(entry);
      }
    }

    return ImmutableList.copyOf(subList);
  }

  @Override
  public CommitResult<Entry> commit(final long index) {
    Preconditions.checkState(START_INDEX <= index && index < mIndex);
    final long oldCommit = mCommitIndex;
    mCommitIndex = index;

    final List<Entry> entries = read(oldCommit + 1, index);

    return new CommitResult<>(entries, index);
  }

  @Override
  public void revert(final long index) {
    Preconditions.checkState(mCommitIndex < index);
    /* delete up to index */
    if (index <= mIndex) {
      while (index < mIndex) {
        mLogStore.remove((int) --mIndex);
      }
    }
  }

  @Override
  public long getCommitIndex() {
    return mCommitIndex;
  }

  @Override
  public long getLastIndex() {
    return mIndex - 1;
  }

  @Override
  public String toString() {
    final List<Entry> committed = new ArrayList<>();
    for (int i = START_INDEX; i <= mCommitIndex; ++i) {
      committed.add(mLogStore.get(i));
    }

    final StringBuilder builder = new StringBuilder().append("Log[ ");
    mLogStore.stream().forEach(entry -> builder.append(" " + entry.getTerm() + " "));
    builder.append(" ]");
    return builder.toString();
  }

  @Override
  public Entry getLastEntry() {
    return read(getLastIndex());
  }

  private void insertOrReplace(final int index, final Entry entry) {
    try {
      if (mLogStore.get(index) != null) {
        mLogStore.remove(index);
      }
    } catch (final IndexOutOfBoundsException e) { }
    mLogStore.add(index, entry);
  }
}

File: core/src/main/java/org/elkd/core/log/InMemoryLog.kt
package org.elkd.core.log

import com.google.common.base.Preconditions
import com.google.common.collect.ImmutableList
import org.elkd.core.consensus.messages.Entry
import java.util.*

class InMemoryLog<E : LogEntry> : Log<E> {
  private val mLogStore: MutableList<E>
  private var mIndex: Long = 0
  override var commitIndex: Long = 0

  override val lastIndex: Long
    get() = mIndex - 1

  override val lastEntry: E
    get() = read(lastIndex)!!

  init {
    mLogStore = ArrayList()

    /* Default Log Record - prevents having to provide a log index that
       technically does not exist e.g index = -1 */
    append(Entry.NULL_ENTRY as E)
  }

  override fun append(entry: E): Long {
    Preconditions.checkNotNull(entry, "entry")

    insertOrReplace(mIndex.toInt(), entry)

    return mIndex++
  }

  override fun append(index: Long, entry: E): Long {
    Preconditions.checkState(0 <= index && index <= mIndex, "index")

    insertOrReplace(index.toInt(), entry)
    return index
  }

  override fun read(index: Long): E? {
    try {
      Preconditions.checkState(START_INDEX <= index && index <= mIndex)
      return mLogStore[index.toInt()]
    } catch (e: Exception) {
      return null
    }

  }

  override fun read(from: Long, to: Long): List<E> {
    Preconditions.checkState(START_INDEX <= from && from <= to)

    val subList = ArrayList<E>()
    for (i in from.toInt()..to) {
      val entry = read(i)
      if (entry != null) {
        subList.add(entry)
      }
    }

    return ImmutableList.copyOf(subList)
  }

  override fun commit(index: Long): CommitResult<E> {
    Preconditions.checkState(START_INDEX <= index && index < mIndex)
    val oldCommit = commitIndex
    commitIndex = index

    val entries = read(oldCommit + 1, index)

    return CommitResult(entries, index)
  }

  override fun revert(index: Long) {
    Preconditions.checkState(commitIndex < index)
    /* delete up to index */
    if (index <= mIndex) {
      while (index < mIndex) {
        mLogStore.removeAt((--mIndex).toInt())
      }
    }
  }

  override fun toString(): String {
    val committed = ArrayList<LogEntry>()
    for (i in START_INDEX..commitIndex) {
      committed.add(mLogStore.get(i.toInt()))
    }

    val builder = StringBuilder().append("Log[ ")
    mLogStore.stream().forEach { entry -> builder.append(" " + entry.term + " ") }
    builder.append(" ]")
    return builder.toString()
  }

  private fun insertOrReplace(index: Int, entry: E) {
    try {
      if (mLogStore[index] != null) {
        mLogStore.removeAt(index)
      }
    } catch (ignored: IndexOutOfBoundsException) {
    }

    mLogStore.add(index, entry)
  }

  companion object {
    private val START_INDEX = 0
  }
}
